File: console/src/lib/api/prompts.ts
import type { Event } from "@zuihitsu/wire/types/Event.ts";
import type { PromptTemplateName } from "@zuihitsu/wire/types/PromptTemplateName.ts";
import type { TemplateStatus } from "@zuihitsu/wire/types/TemplateStatus.ts";
import type { LiveConnection } from "./live.ts";
import { authHeaders, errorMessage } from "./http.ts";

export type { TemplateStatus };

/// One prompt template at its current version — the highest version registered under a name (spec
/// §Initialization → prompt templates). The agent reads the system-prompt scaffold and the framing
/// templates from these.
export interface PromptTemplate {
  name: PromptTemplateName;
  version: number;
  body: string;
}

/// The current prompt templates up to the cursor: the highest-versioned `PromptTemplateRegistered`
/// per name, in registration order. Templates are read from the log, never materialized, so the
/// console derives them the same way the agent does.
export function deriveTemplates(events: Event[], cursor: number): PromptTemplate[] {
  const latest = new Map<PromptTemplateName, PromptTemplate>();
  for (const event of events) {
    if (event.seq > cursor) continue;
    const payload = event.payload;
    if (payload.type !== "PromptTemplateRegistered") continue;
    const current = latest.get(payload.name);
    if (!current || payload.version >= current.version) {
      latest.set(payload.name, {
        name: payload.name,
        version: payload.version,
        body: payload.body,
      });
    }
  }
  return [...latest.values()];
}

/// Each template name's status against the running build's defaults (`GET /control/prompt-status`):
/// whether it is a curated (operator-edited) surface, the build's newest default version, and whether a
/// newer default is available for a curated surface to adopt. The console badges the curated surfaces
/// carrying a pending upgrade; a default-tracking name auto-tracks the build at boot, so it never
/// reports one. The build defaults live in Rust, so this is fetched rather than derived from events.
export async function getTemplateStatuses(connection: LiveConnection): Promise<TemplateStatus[]> {
  const response = await fetch(`${connection.baseUrl}/control/prompt-status`, {
    headers: authHeaders(connection),
  });
  if (!response.ok) throw new Error(await errorMessage(response));
  return (await response.json()) as TemplateStatus[];
}

/// Register a new version of a prompt template — the operator edit. The new version replaces the
/// template from the next read on, and arrives back through the live tail like any other event.
export async function registerPrompt(
  connection: LiveConnection,
  name: PromptTemplateName,
  body: string,
): Promise<void> {
  const response = await fetch(`${connection.baseUrl}/control/prompt`, {
    method: "POST",
    headers: authHeaders(connection),
    body: JSON.stringify({ name, body }),
  });
  if (!response.ok) throw new Error(await errorMessage(response));
}

File: crates/frontend-types/src/prompts.rs
//! The prompt-template status the console reads (`GET /control/prompt-status`): each build-default
//! template name's registration state against the running build's defaults. Crosses the wire to the
//! console, so it lives here; the main crate computes it from the log and its `default_templates`.

use serde::Serialize;
use zuihitsu_core::event::PromptTemplateName;

/// One prompt template name's status against the build defaults. The console badges a curated surface
/// with a newer default available; the doctrine is that a default-tracking name (Bootstrap-latest)
/// auto-tracks the build at boot, while a curated name (operator-edited) is sovereign and adopts a new
/// default only on the operator's explicit choice.
#[derive(Clone, Debug, PartialEq, Serialize)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
pub struct TemplateStatus {
    pub name: PromptTemplateName,
    /// The highest version registered on the log for this name (`0` if the name is somehow absent —
    /// which a booted agent's reconcile has already backfilled away).
    pub latest_version: u32,
    /// The latest registration is operator-sourced: a curated surface the boot reconcile never
    /// auto-touches. A default-tracking name is `false`.
    pub curated: bool,
    /// The build's newest default version for this name.
    pub default_version: u32,
    /// A newer or divergent build default is available for a curated surface to adopt. Always `false`
    /// for a default-tracking name, which the boot reconcile has already advanced to the build default.
    pub upgrade_available: bool,
}
